#!/bin/bash
#SBATCH --job-name=chromHMM_LearnModel.mm39
#SBATCH --output=/home/mnt/network/DATA/projects/isgs/analysis/chromHMM/reports/chromHMM_LearnModel.mm39_%A.out
#SBATCH --time=48:00:00
#SBATCH --partition=hmem
#SBATCH --cpus-per-task=20
#SBATCH --qos=qos_batch

######################################################
# README

# this script runs ChromHMM LearnModel for files in a specified directory, sent to a specified queue
	# pass x1 command line argument to run this script
		# the number of states to generate in the model
# e.g., to generate 21 state model, run:
# sbatch /path/to/script/ 21

######################################################
# MODULES + PATHS + REPORTING

## Load modules
module load openjdk 	## load java module

## Report relevant parameters
echo "~ChromHMM parameters and info~"

echo "genome: GRCm39/mm39, annotation: Ensembl v110 (gtf), running with custom assembly files for chromHMM ( assembly: custom_mm39_ensembl )"
echo "running ChromHMM LearnModel on bam files generated by nfcore chipseq and atacseq piplines"
echo "binsize is 200 (default) and shiftsize is 100 (default)"

## Specify paths and information
inBin="/home/mnt/network/DATA/projects/isgs/analysis/chromHMM/ptm_tar/binarized_bam/targets"
out_res="/home/mnt/network/DATA/projects/isgs/analysis/chromHMM/ptm_tar/models"

genome="/home/genomes/mm39/GRCm39_primary_assembly.fa.gz"
chrom_sz="/home/genomes/mm39/GRCm39_primary_assembly.chrom_sizes"

######################################################
# RUN

java --version

## Report number of states
echo "Generating a model with $1 states"

echo "Running LearnModel"
java -mx8000M -jar /home/programs/ChromHMM/ChromHMM.jar LearnModel -b 200 -i "chrHMM.new_ptm_tar" -l ${chrom_sz} -noautoopen -p 0 ${inBin} ${out_res} $1 custom_mm39_ensembl

echo "Run completed"
